[Week1][Davidyang2149] Create ToDo REST API - Done

DavidYang2149 opened the pull request:
# 과제 - ToDo REST API 만들기
## 업데이트 내역
- [x] 2022.03.21 - ~07:00 AM / 최초 PR 작성
- [x] 2022.03.21 - ~11:26 PM / 1강 강의 완독 후 아샬님의 작성하신 코드를 따라서 연습
- [x] 2022.03.22 - ~11:57 PM / 월요일에 작성한 코드를 개선하고, ToDo 상세 조회하기를 구현
- [x] 2022.03.23 - ~ 11:30 PM /  ToDo 제목 수정하기, ToDo 삭제하기를 구현 - 과제 목표 달성

## 과제 목표
- [x] ToDo 목록 얻기 - GET /tasks
- [x] ToDo 상세 조회하기 - GET /tasks/{id}
- [x] ToDo 생성하기 - POST /tasks
- [x] ToDo 제목 수정하기 - PUT/PATCH /tasks/{id}
- [x] ToDo 삭제하기 - DELETE /tasks/{id}

## 요구사항
- [ ] 모든 API 테스트를 통과해야 합니다.

## 작업내용
- [x] 1강 강의 완독
- [x] 과제 목표 달성하기

## 궁금한 점
- [x] 자바의 버전 관리 (스프링 5기 채널에 질문 후 답변 받음)
- [x] REST API에 대하여 (스프링 5기 채널에 질문 후 답변 받음)  

## 배운 것과 조사할 것
- [x] REST API란 무엇인지 다른 사람에게 설명해보기
- [x] 이번 강의에서 다루지 않을 HATEOAS 조사하기

DavidYang2149 committed:
아무것도 동작하지 않는 HttpServer를 제작합니다

- 일등으로 PR을 올리기 위해 아무것도 동작하지 않는 HttpServer를 제작해봅니다.

DavidYang2149 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,22 @@
     }
 
     public static void main(String[] args) {
+        InetSocketAddress address = new InetSocketAddress(8000);
+        try {
+            HttpServer httpServer = HttpServer.create(address, 0);
+            httpServer.start();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
일등으로 PR하기 위해 아무것도 동작하지 않는 HttpServer를 생성합니다. 😄

ahastudio replied:
👍

johngrib replied:
👍

gyedongjeon replied:
👍

hannut91 replied:
👍

DavidYang2149 committed:
불완전한 tasks 통신을 구축합니다

- 1강 강의 완강을 인증하기 위해 아샬님의 코드를 따라 합니다.
- 어딘가 개선이 필요하다고 생각되는 "GET /tasks"를 만들었습니다.
- 어딘가 개선이 필요하다고 생각되는 "POST /tasks"를 만들었습니다.
- TODO: 내일(화)부터 남은 과제 목표의 기능을 구현합니다.
- TODO: 금일(월) 작성한 코드에서 개선할 수 있는지 코멘트를 남깁니다.

DavidYang2149 commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,76 @@
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+
+import java.io.*;
```
## 훔치기
### Facts (사실, 객관)
- `import *`를 지양해야 한다는 코멘트를 보았습니다.  
### Feelings (느낌, 주관)
- 어떤 것을 `import` 정확히 명시되어야 좋겠다는 근거가 있었습니다.
### Findings (배운점) 및 Affirmation(자기 선언)
- 자동 import를 사용하니 무심하게 지나쳤던 부분이었습니다.
- 종립님의 글에서 [IntelliJ 설정 방법](https://johngrib.github.io/wiki/intellij/#import--%EB%B0%A9%EC%A7%80%ED%95%98%EA%B8%B0)을 참조하여 설정을 변경했습니다.

DavidYang2149 replied:
DONE: https://github.com/CodeSoom/spring-week1-assignment-1/pull/74/commits/29b6b00f9fba4e939084eb33accb8a4315dc777e

johngrib replied:
👏

DavidYang2149 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
 
     public static void main(String[] args) {
+        try {
+            InetSocketAddress address = new InetSocketAddress(8000);
```
## 훔치기
### Facts (사실, 객관)
- `Magic literal`을 상수로 추출할 수 있다는 코멘트를 보았습니다.
### Feelings (느낌, 주관)
- 코드를 읽는 사람이 해당 숫자(8000)의 의미를 명확하게 이해할 수 있도록, 가독성을 높이는 목적을 가지고 있다고 했습니다.
### Findings (배운점) 및 Affirmation(자기 선언)
- TODO: 해당 숫자를 상수로 추출하려 합니다.

DavidYang2149 replied:
DONE: https://github.com/CodeSoom/spring-week1-assignment-1/pull/74/commits/0d56617275111988815809061aa7fd2889e84873

DavidYang2149 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
+import java.net.InetSocketAddress;
+
 public class App {
     public String getGreeting() {
         return "Hello World!";
     }
```
TODO: 해당 메서드가 사용되지 않아 삭제하고자 합니다.

DavidYang2149 replied:
DONE: https://github.com/CodeSoom/spring-week1-assignment-1/pull/74/commits/cda3b4901cd1758c80d96f4f2f9917f2287fdf00

DavidYang2149 commented:
# 2022.03.21 Mon
## Facts (사실, 객관)
- '스프링 1강의 웹개발 시작하기' 강의를 완강했습니다.
- 1주차 과제 참여자 중 일등으로 PR을 올렸습니다.
- 아샬님의 강의를 토대로 코드를 따라 작성했습니다.
## Feelings (느낌, 주관)
- '완벽'보다는 '빠르고 자주 피드백'을 받는 것에 중점을 두겠습니다.
- 오늘처럼 데일리 회고를 자주 작성해서 주간회고를 더 쉽고 빠르게 작성하려 합니다.
## Findings (배운점) 및 Affirmation(자기 선언)
- HTTPie를 사용하여 쉽고 빠르게 API를 테스트하는 방법을 배웠습니다.
- '모든 것'을 성공하려 하지 말고 '할 수 있는 것'을 성공시키겠습니다.
## 고민 (메모)
- 내일(화요일) 과제를 진행하기 위해 집중할 수 있는 일이 두 가지가 있습니다.
  - 첫 번째는 오늘 만든 GET, POST를 개선하는 일입니다.
  - 두 번째는 오늘 구현하지 못한 잔여 기능을 구현하는 일입니다.
- 첫 번째를 먼저 진행하면 개선된 코드를 기준으로 잔여 기능을 더 쉽게 구현할 수 있을 것 같습니다.
- 두 번쨰를 먼저 진행하면 전체적인 흐름을 확인하고 코드 개선에 더 도움이 될 것 같습니다.
- ~~어떤 방법으로 진행할지 고민 중입니다.~~ / 짧게 코드를 개선해보고 잔여 기능을 구현하기로 결정했습니다!

DavidYang2149 committed:
App에 있었던 인사 메서드를 치웁니다

- App은 처음에 인사만 할 수 있었으나, 지금은 tasks를 관리하는 API 역할을 합니다.
- 더 이상 App이 인사할 필요가 없기 때문에 환영 메서드를 정리합니다.

DavidYang2149 committed:
매직 리터럴을 상수로 추출합니다

- 매직 리터럴을 상수로 추출하여 명확한 의미를 전달합니다.
- https://refactoring.com/catalog/replaceMagicLiteral.html
- https://www.youtube.com/watch?v=cWRwAaUM1Mc

DavidYang2149 committed:
아낌없이 주지 않고 필요한 것만 수입하도록 변경합니다

- import * 를 사용을 지양하고 무엇을 import 받는지 명확하게 표현하도록 합니다.

DavidYang2149 commented:
## 코드 개선
- 월요일에 작성한 코드를 개선하였습니다.
- 더 개선하고 싶은 욕심이 있지만, 우선 여기까지만 코드를 개선하겠습니다.
- 과제 목표를 먼저 달성 후 코드를 개선하려고 합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,81 @@
+
+public class TodoHttpHandler implements HttpHandler {
+    private List<Task> tasks = new ArrayList<>();
+    private ObjectMapper objectMapper = new ObjectMapper();
```
이번 1강에서는 `ObjectMapper`를 사용하지 않고 과제를 수행해 보도록 합시다. (나중에 강의에서 나오게 되는데 그때부터 사용해 봅시다)

johngrib replied:
오 이건 제가 강의 과정을 착각했어요. 3강에서 나오는 줄 알았는데 1강부터 나오네요. 지금도 사용하셔도 됩니다. 죄송합니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,81 @@
+
+        String content = "Hello, World!";
+
+        if (method.equals("GET") && path.equals("/tasks")) {
```
주소가 `/tasks/`로 들어오면 어떻게 될까요?

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,81 @@
+            content = tasksToJSON();
+        }
+
+        if (method.equals("POST") && path.equals("/tasks")) {
```
`GET`, `POST` 등은 enum으로 따로 정의하기 좋게 생겼군요. 한번 해보면 어떨까요?

johngrib reviewed (commented):
훌륭한 전략입니다!

DavidYang2149 committed:
잘못된 경계 밖에 있는 명령문을 안으로 넣어줍니다

- 테스크 생성 명령문이 잘못된 위치에 있기에, 올바른 위치로 옮깁니다.

DavidYang2149 committed:
어딘가 부족해 보이는 테스크 추가 기능을 보완합니다

- 테스크를 추가할 때 id를 순차적으로 +1씩 증가시킵니다.
- 기능을 개선하고 싶은 욕구가 있지만 일단 참습니다.

DavidYang2149 committed:
어딘가 부족해 보이는 테스크 상세보기를 구현합니다

- 테스크 상세보기를 구현하였습니다.
- TODO: 테스크 id가 존재하지 않을 경우에 대한 대응이 필요합니다.

DavidYang2149 commented:
# 2022.03.22 Tue

## Facts (사실, 객관)

- 1주차 과제에서 월요일에 작성한 코드 중 세 개의 주제를 개선했습니다.
- 1주차 과제에서 ToDo 생성하기 기능을 보완했습니다.
- 1주차 과제에서 ToDo 상세보기 기능을 구현했습니다.

## Feelings (느낌, 주관)

- 오늘 마음 속으로는 과제 목표를 모두 달성하려 했는데 실패하였습니다.
- 어디서 시간이 오래 걸렸는지 명확히 판단하지 못했습니다.
- 자바스크립트를 작성할 때와는 다르게 자바는 형변환에 대해 신경쓸 부분이 많았습니다.

## Findings (배운점) 및 Affirmation(자기 선언)

- 오래 걸린 것에서 무엇을 배울 수 있었는지를 기록하려 합니다.
- 앞으로 작업을 시작하기 전에 오늘 어떤 목표를 달성할지 먼저 정리하려고 합니다.
- 앞으로 작업을 진행하면서 30분 마다 상태를 점검하고 목표를 달성할 수 있는지 진단합니다.

## Memo (메모)
- 종립님의 피드백을 잘 기록해서 코드를 개선하려 합니다.
- 자바의 List filter 기능을 적용하는데 시간이 많이 소요되었습니다.
- javadoc를 읽어도 머리 속에 잘 이해되지 않습니다.
- javadoc가 익숙해질 때까지 읽고 이해하는 연습을 해야겠습니다.

DavidYang2149 commented:
## 오늘 목표 설정하기 (화요일)
- ToDo 제목 수정하기 - PUT/PATCH /tasks/{id} / 구현하기
- ToDo 삭제하기 - DELETE /tasks/{id} / 구현하기
- 전략: 2개의 목표를 각각 1시간 이내로 구현하기
- 오늘까지 빠르게 목표를 달성하려고 합니다.
- 내일부터 코드를 개선하는 작업을 시작하고 싶습니다.
- 내일부터 부족한 기능과 예외를 다듬는 작업을 시작하고 싶습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,100 @@
+            content = tasksToJSON();
+        }
+
+        if (method.equals("GET") && path.split("/")[1].equals("tasks") && path.split("/").length > 2) {
```
`method.equals("GET")` 방식의 코드는 조심해야 하는 점이 있습니다. `method`가 `null`이면 여기에서 NPE가 터지거든요. 그래서 숙련된 자바 애플리케이션 개발자들은 `"GET".equals(method)` 이런 방식의 코드를 작성합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,100 @@
+            System.out.println(task);
+        }
+
+        exchange.sendResponseHeaders(200, content.getBytes().length);
```
`200`도 enum으로 정의하기 좋아 보입니다.

johngrib commented:
JavaDoc은 `String.java`를 열어서 읽어보시는 것부터 시작해보세요. 좋은 공부가 될 거에요.

DavidYang2149 committed:
어딘가 부족해 보이는 테스크 제목 수정하기를 구현합니다

- PUT에 대응하는 제목 수정하기를 구현 하였습니다.
- PATCH에 대응하는 제목 수정하기를 구현 하였습니다.

DavidYang2149 committed:
어딘가 부족해 보이는 테스크 삭제하기를 구현합니다.

- 테스크 삭제하기를 구현 하였습니다.
- 과제 목표를 모두 달성했습니다.

DavidYang2149 commented:
## 오늘 목표 결과 (화요일)
- [x] ToDo 제목 수정하기 - PUT/PATCH /tasks/{id} / 구현하기 - 성공
- [x] ToDo 삭제하기 - DELETE /tasks/{id} / 구현하기 - 성공
- [x] (전략) 2개의 목표를 각각 1시간 이내로 구현하기 - 성공
- 오늘까지 과제 목표를 달성했습니다.
- 내일부터 코드를 개선하고, 부족한 기능을 다듬으며, 예외를 처리하는 작업을 시작할 예정입니다!

DavidYang2149 commented:
# 2022.03.23 Wed

## Facts (사실, 객관)

- 1주차 과제에서 ToDo 수정하기 기능을 구현했습니다.
- 1주차 과제에서 ToDo 삭제하기 기능을 구현했습니다.
- 1주차 과제 목표를 달성했습니다.
- 코드숨 디스코드에 REST API에 대한 질문을 남겼습니다.

## Feelings (느낌, 주관)

- 오늘 과제 목표를 모두 달성해서 기분이 좋습니다.
- 시작 전 오늘의 목표를 설정했고, 이를 지켰습니다.
- 질문을 던졌는데 제가 궁금한 부분이 사실은 빙산의 일각이고, 그 뿌리는 더 깊었다는 것을 알게 되었습니다.
- 고민을 적게 하는 방법을 고민해보려고 합니다.

## Findings (배운점) 및 Affirmation(자기 선언)

- 학습을 할 때 개념의 기원에서 가장 가까운 곳에서부터 시작해야겠습니다.
- 내일부터는 요구사항을 충족시킬 수 있도록 코드를 개선하려고 합니다.
- 내일은 종립님의 추천에 따라 javadoc 중 String.java 읽고 정리해보려고 합니다.

## Memo (메모)

- 드디어 내일부터 코드를 개선하려고 합니다.
- 종립님의 피드백을 통해 코드의 품질을 더 좋게 만들겠습니다.

DavidYang2149 commented:
## 오늘 목표 설정하기 (수요일)
- 금일은 육아로 정신이 지쳐서 목표를 작게 잡았습니다.
- `String.java` 읽기
- 종립님이 주신 피드백을 토대로 다른 분들의 코드 훔치기

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,134 @@
+            Long id = Long.valueOf(path.split("/")[2]);
+
+            content = taskToRemove(id);
+        }
```
공통적으로 사용하고 있는 코드 조각들이 보이는군요. `path.split("/")`, `length > 2`
일단 빠른 구현을 위해 이렇게 작업하신 거겠죠? 여유가 생기면 DRY 원칙을 실천해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -0,0 +1,134 @@
+        outputStream.close();
+    }
+
+    private String taskToRemove(Long id) {
+        tasks.removeIf(task -> task.getId().equals(id));
+        return "Delete a task.";
+    }
```
이 메소드의 리턴값의 의미가 불분명해보입니다. 마치 로그로 남겨야 하는 문장을 리턴하고 있는 것처럼 보여요. `private` 메소드이긴 하지만 JavaDoc 주석을 작성해서 동료들이 이해할 수 있도록 해보세요. 그러면 뭔가가 더 보이게 될 겁니다.

DavidYang2149 commented:
# 2022.03.24 Thu

## Facts (사실, 객관)

- 1주차 과제에서 다른 분들이 작성하신 코드를 살펴보았습니다.
- String javadoc를 읽었습니다.

## Feelings (느낌, 주관)

- 육아로 정신이 지쳐서 금일은 빠르게 쉬려고 합니다.
- 그냥 자기가 아까운 마음에 내일 무엇을 하고 싶은지 눈으로 보고 생각을 남깁니다.

## Findings (배운점) 및 Affirmation(자기 선언)

- 8주 동안 장거리 달리기를 위해 페이스를 조절하고자 합니다.
- 컨디션이 좋지 않은 날도 있습니다. 현재 제 기분을 인지하고 빠르게 회복하는데 목표를 두겠습니다.

## Memo (메모)

- 내일 훔치기를 통해 코드를 빠르게 개선해보려고 합니다.

DavidYang2149 commented:
## 오늘 목표 설정하기 (금요일)
- 이번주가 거의 끝나가므로 가장 작고 빠르게 진행할 수 있는 코드부터 개선하겠습니다.
- 눈에 보이는 것부터 빠르게 처리하고 이곳에 하나씩 내용을 추가하겠습니다.
- GET, POST 등을 enum으로 정의하기
- 200 등을 enum으로 정의하기

DavidYang2149 committed:
HTTP 메서드를 정의합니다

- HTTP 메서드를 enum으로 추출합니다.
- 추출한 enum을 넣을 수 있는 코드에 적용합니다.

DavidYang2149 committed:
HttpMethod의 코드를 개선합니다

- 열거형 HttpMethod로 변환할 수 있는 메서드를 제작합니다.
- 개선할 수 있는 코드에 setHttpMethod를 적용합니다.

DavidYang2149 committed:
HTTP 상태를 정의합니다

- HTTP Status를 enum으로 추출합니다.
- 추출한 enum을 넣을 수 있는 코드에 적용합니다.

DavidYang2149 commented:
# 2022.03.25 Fri

## Facts (사실, 객관)

- 1주차 과제에서 열거형을 이용하여 코드를 개선하였습니다.
- 이펙티브 자바 중 열거형에 대한 내용을 읽었습니다.
- RFC 7231 문서의 일부를 읽었습니다.

## Feelings (느낌, 주관)

- 문서 읽기가 어려웠습니다. 항상 쉬운 내용을 찾아서 읽었다는 것을 깨달았습니다.
- 공식 문서를 읽고 이해하는 방법을 찾아봐야겠습니다.

## Findings (배운점) 및 Affirmation(자기 선언)

- 문서 읽기를 습관화하기 위해 매일 반복해서 봐야겠습니다.
- 작지만 꾸준한 습관을 만들어 가려고 합니다.

## Memo (메모)

- 종립님의 ‘나만의 위키를 만들어보자'를 보았는데 흥미를 느껴서 저도 시도해보려고 합니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -48,17 +49,17 @@ public void handle(HttpExchange exchange) throws IOException {
         String content = "Hello, World!";
 
-        if (method.equals("GET") && path.equals("/tasks")) {
+        if (HttpMethod.GET.toString().equals(method) && path.equals("/tasks")) {
```
static import 하는 방법을 조사해서 `HttpMethod.GET`에 사용해 보세요.

DavidYang2149 replied:
static import를 조사하고 방법을 적용해보았습니다.
- 참조: https://docs.oracle.com/javase/specs/jls/se11/html/jls-7.html#jls-7.5.3
- 참조: https://jcp.org/aboutJava/communityprocess/jsr/tiger/static-import.html
DONE: https://github.com/CodeSoom/spring-week1-assignment-1/pull/74/commits/004188a63ad7edcdcd5b89b78d83ea3dd7eeb0a7

johngrib commented on app/src/main/java/com/codesoom/assignment/TodoHttpHandler.java:
```diff
@@ -49,17 +49,17 @@ public void handle(HttpExchange exchange) throws IOException {
         }
 
-        if (HttpMethod.GET.toString().equals(method) && path.split("/")[1].equals("tasks") && path.split("/").length > 2) {
+        if (HttpMethod.GET.equals(method) && path.split("/")[1].equals("tasks") && path.split("/").length > 2) {
```
if 문 안쪽이 꽤 복잡합니다. 잘 설명해주는 이름을 가진 private 메소드로 추출해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/HttpMethod.java:
```diff
@@ -0,0 +1,26 @@
+public enum HttpMethod {
+    GET, POST, PUT, PATCH, DELETE;
+
+    public static HttpMethod setHttpMethod(String method) {
```
이런 경우엔 이런 이름을 써보세요.
```suggestion
    public static HttpMethod valueOf(String method) {
```

DavidYang2149 replied:
DONE: https://github.com/CodeSoom/spring-week1-assignment-1/pull/74/commits/053c8fe7fd621c4323cce86e1bb81a02d021b350

DavidYang2149 committed:
import static을 이용하여 통신 명세서를 명료하게 표현합니다

- 참조: https://docs.oracle.com/javase/specs/jls/se11/html/jls-7.html#jls-7.5.3
- 참조: https://jcp.org/aboutJava/communityprocess/jsr/tiger/static-import.html

DavidYang2149 committed:
setHttpMethod를 대체할 수 있는 valueOf 메서드를 적용합니다

DavidYang2149 committed:
훔치기 javadoc @see를 적용합니다

- 참조: https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835233263

johngrib reviewed (approved):
한 주간 고생 많으셨습니다!

DavidYang2149 commented:
한 주간 많은 가르침 감사합니다!
1주차 주간 회고를 작성하였습니다.
https://everlasting-jar-961.notion.site/2022-03-27-Sun-26c173f5bda449f4a6fc1d99f720bc07
